runtime/lua/vim/lsp/document_color.lua

```diff
@@ -297,6 +297,31 @@ function M.enable(enable, bufnr, opts)
   end
 end
 
+api.nvim_create_autocmd('LspAttach', {
```

Comment:
I don't think this is the best way to hook this up. Right now the recommended `documentColor` setup is to have something like:
```lua
vim.api.nvim_create_autocmd('LspAttach', {
  callback = function(args)
    local client = vim.lsp.get_client_by_id(args.data.client_id)
    if client:supports_method('textDocument/documentColor') then
      vim.lsp.document_color.enable(true, args.buf)
    end
  end
})
```

With dynamic registration `supports_method` would return false, and if `enable` is never called for that buffer then we could miss a client that later registers.

Should we maybe change this API to be called in `BufEnter` autocommand instead of on `LspAttach`?

Reply:
Open to other ideas of course.

Reply:
do we have any way of listening to a "dynamic registration" event/callback ?

Reply:
AFAIK we don't (until this PR). Here I introduce the private callback mechanism, but I'm open to adding a new `LspRegistration` or `LspCapability` autocommand instead.

Reply:
In my config, I'm hooking into`client/registerCapability` directly to do things like this https://github.com/TheLeoP/nvim-config/blob/4cbf283a59186e14ebf3139744ba8c995b14fcae/lua/personal/config/lsp.lua#L113 . Just like it's done for lsp_defaults. Is this any different?

Actually, I may have copied this from https://github.com/MariaSolOs/dotfiles/blob/36353ae6dac3887ddefa8d1dd3f3f878cc25f0c9/.config/nvim/lua/lsp.lua#L223

Reply:
> hooking into`client/registerCapability` directly to do things like this https://github.com/TheLeoP/nvim-config/blob/4cbf283a59186e14ebf3139744ba8c995b14fcae/lua/personal/config/lsp.lua#L113

oh yeah for sever-to-client events that is the recommended, documented way.

do we need to augment `vim.lsp.handlers` in this PR, instead of LspAttach ?

Reply:
> do we need to augment `vim.lsp.handlers` in this PR, instead of LspAttach/BufEnter ?

The thing is that then we would have a bunch of `if` blocks in the registration handler if we do this, and instead I find it cleaner to have the specific LSP modules hook up to registration events if they need to.

Reply:
It would save us needing to worry about LspAttach vs BufEnter, right? A flat list of if-else doesn't sound too bad.

Reply:
Okay yeah thinking about it it doesn't feel too bad to move the logic to `vim.lsp.handlers`. We still have the issue of when it makes the most sense to call `vim.lsp.document_color.enable`:
- Continue using the documented `LspAttach` approach but remove the check for `supports_method` as a client could dynamically register later.
- Since `LspAttach` could be invoked several times for the same buffer, suggest using `BufEnter` instead (`document_color.enable` doesn't take a `client_id` parameter).

Reply:
Alright I went with the first approach in https://github.com/neovim/neovim/compare/7e31180682bc7a08c36449a6cfdbfa9622f23b27..ed3dbce52410b8b524b5ca13842801bfd25a1b47, but I'm open to better ideas.